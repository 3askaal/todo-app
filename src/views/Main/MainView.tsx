import React, { useState, useEffect, FC, useReducer, FormEvent } from 'react'
import { useMutation, useLazyQuery } from 'react-apollo'
import { useDebouncedCallback } from 'use-debounce/lib'
import { Plus as PlusIcon } from 'react-feather'
import { some } from 'lodash'
import { TodosQuery } from 'queries'
import { Todo, Container, Spacer, Input, Button } from 'components'
import { TodoReducer, initialTodoState } from 'context'
import {
  CheckAllMutation,
  UncheckAllMutation,
  DeleteAllMutation,
  CreateMutation,
  DeleteMutation,
  UpdateMutation,
} from '../../mutations'
import { SIndexViewCreate, SIndexViewError, SIndexViewOptions } from './MainView.styled'
import { TTodo, TMutation, TTodoInput } from '../../../../shared/types/todo.d'

export const MainView: FC<{}> = () => {
  const [runTodosQuery, { data: { Todos = [] } = {} }] = useLazyQuery(TodosQuery)
  const [todos, dispatch] = useReducer(TodoReducer, initialTodoState)
  const [newContent, setNewContent] = useState<string>('')
  const [error, setError] = useState<string | null>(null)
  const [createMutation, { data: createData }] = useMutation<TMutation>(CreateMutation)
  const [updateTodoMutation] = useMutation<TMutation>(UpdateMutation)
  const [deleteTodoMutation] = useMutation<TMutation>(DeleteMutation)
  const [checkAllMutation] = useMutation<TMutation>(CheckAllMutation)
  const [uncheckAllMutation] = useMutation<TMutation>(UncheckAllMutation)
  const [deleteAllMutation] = useMutation<TMutation>(DeleteAllMutation)

  useEffect(() => {
    runTodosQuery()
  }, [runTodosQuery])

  useEffect(() => {
    if (Todos?.length) {
      dispatch({ type: 'set', value: Todos })
    }
  }, [Todos])

  useEffect(() => {
    if (createData?.createTodo) {
      dispatch({ type: 'set', value: [...todos, createData.createTodo] })
    }
  }, [createData])

  useEffect(() => {
    if (error) {
      setError(null)
    }
  }, [newContent])

  function onCreate(event: FormEvent): void {
    event.preventDefault()

    if (!newContent) {
      setError('Content field can not be empty')
      return
    }

    createMutation({
      variables: {
        data: {
          content: newContent,
        },
      },
    })

    setNewContent('')
  }

  const [debouncedUpdate] = useDebouncedCallback(async (_id: string, update: TTodoInput) => {
    await updateTodoMutation({
      variables: {
        _id,
        data: update,
      },
    })
  }, 500)

  async function onUpdate(_id: string, update: TTodoInput): Promise<void> {
    dispatch({ type: 'update:one', _id, update })
    await debouncedUpdate(_id, update)
  }

  async function onDelete(_id: string): Promise<void> {
    dispatch({ type: 'delete:one', _id })
    await deleteTodoMutation({
      variables: {
        _id,
      },
    })
  }

  async function onCheckAll(): Promise<void> {
    dispatch({ type: 'check:all' })
    await checkAllMutation()
  }

  async function onUncheckAll(): Promise<void> {
    dispatch({ type: 'uncheck:all' })
    await uncheckAllMutation()
  }

  async function onDeleteAll(): Promise<void> {
    dispatch({ type: 'delete:all' })
    await deleteAllMutation()
  }

  return (
    <Container>
      <Spacer>
        <SIndexViewCreate data-testid="main-create">
          <form onSubmit={onCreate} data-testid="main-create-form">
            <Input
              value={newContent}
              onChange={event => setNewContent(event.target.value)}
              error={error}
              data-testid="main-create-form-input"
            />
          </form>
          <Button onClick={onCreate} data-testid="main-create-button">
            <PlusIcon />
          </Button>
        </SIndexViewCreate>
        {error ? <SIndexViewError data-testid="main-create-error">{error}</SIndexViewError> : null}
        <Spacer size="s" data-testid="main-list">
          {todos.map(({ _id, content, completed }: TTodo) => (
            <Todo
              content={content ?? ''}
              completed={completed ?? false}
              key={_id}
              onUpdate={(update: TTodoInput) => onUpdate(_id, update)}
              onDelete={() => onDelete(_id)}
            />
          ))}
        </Spacer>
        {todos.length ? (
          <SIndexViewOptions data-testid="main-options">
            {some(todos, { completed: false }) && (
              <Button data-testid="main-options-checkall" onClick={onCheckAll}>
                Check All
              </Button>
            )}
            {some(todos, { completed: true }) && (
              <Button data-testid="main-options-uncheckall" onClick={onUncheckAll}>
                Uncheck All
              </Button>
            )}
            <Button data-testid="main-options-deleteall" onClick={onDeleteAll}>
              Delete All
            </Button>
          </SIndexViewOptions>
        ) : null}
      </Spacer>
    </Container>
  )
}

export default MainView
