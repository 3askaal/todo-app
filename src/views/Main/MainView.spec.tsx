import React from 'react'
import { MainView } from './MainView'
import {
  render,
  act,
  fireEvent,
  customBeforeEach,
  customAfterEach,
  waitForData,
} from '../../test/utils'

let container: any = null

beforeEach(() => {
  container = customBeforeEach()
})

afterEach(() => {
  customAfterEach(container)
})

describe('MainView', () => {
  it('runs', async () => {
    const tree = render(<MainView />)

    await act(async () => {
      await waitForData()
    })

    expect(tree).toMatchSnapshot()
  })

  it('creates todo', async () => {
    const { getByTestId, getAllByTestId } = render(<MainView />)

    await act(async () => {
      await waitForData()
    })

    const createFormInput: any = getByTestId('main-create-form-input')
    fireEvent.change(createFormInput, { target: { value: 'New todo' } })

    const createFormButton: any = getByTestId('main-create-button')
    fireEvent.click(createFormButton)

    await act(async () => {
      await waitForData()
    })

    expect(getAllByTestId('todo').length).toBe(3)
  })

  it('creates todo (fails)', async () => {
    const { getByTestId, queryByTestId } = render(<MainView />)

    await act(async () => {
      await waitForData()
    })

    const createFormButton: any = getByTestId('main-create-button')
    fireEvent.click(createFormButton)

    await act(async () => {
      await waitForData()
    })

    expect(getByTestId('main-create-error')).toBeDefined()

    const createFormInput: any = getByTestId('main-create-form-input')
    fireEvent.change(createFormInput, { target: { value: 'New todo' } })

    expect(queryByTestId('main-create-error')).toBeNull()
  })

  it('updates todo', async () => {
    const { getAllByTestId } = render(<MainView />)

    await act(async () => {
      await waitForData()
    })

    const todoToUpdate: any = getAllByTestId('todo-content-input')[1]
    expect(todoToUpdate.value).toBe('find new job')

    fireEvent.change(todoToUpdate, {
      target: { value: 'find awesome new job' },
    })

    await act(async () => {
      await waitForData(1000)
    })

    expect(todoToUpdate.value).toBe('find awesome new job')
  })

  it('removes todo', async () => {
    const { getAllByTestId } = render(<MainView />)

    await act(async () => {
      await waitForData()
    })

    expect(getAllByTestId('todo').length).toBe(2)

    const todoToRemoveButton: any = getAllByTestId('todo-button')[1]
    fireEvent.click(todoToRemoveButton)

    await act(async () => {
      await waitForData()
    })

    expect(getAllByTestId('todo').length).toBe(1)
  })

  it('check all', async () => {
    const { getByTestId, queryByTestId } = render(<MainView />)

    await act(async () => {
      await waitForData()
    })

    const checkAllButton: any = getByTestId('main-options-checkall')
    fireEvent.click(checkAllButton)

    await act(async () => {
      await waitForData()
    })

    expect(queryByTestId('main-options-checkall')).toBeNull()
  })

  it('uncheck all', async () => {
    const { getByTestId, queryByTestId } = render(<MainView />)

    await act(async () => {
      await waitForData()
    })

    const uncheckAllButton: any = getByTestId('main-options-uncheckall')
    fireEvent.click(uncheckAllButton)

    await act(async () => {
      await waitForData()
    })

    expect(queryByTestId('main-options-uncheckall')).toBeNull()
  })

  it('delete all', async () => {
    const { getByTestId, queryByTestId, queryAllByTestId } = render(
      <MainView />,
    )

    await act(async () => {
      await waitForData()
    })

    const deleteAllButton: any = getByTestId('main-options-deleteall')
    fireEvent.click(deleteAllButton)

    await act(async () => {
      await waitForData()
    })

    expect(queryByTestId('main-options-deleteall')).toBeNull()
    expect(queryAllByTestId('todo').length).toBe(0)
  })
})
