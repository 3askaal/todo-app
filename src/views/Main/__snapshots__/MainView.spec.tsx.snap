// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainView runs 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <div
        class="sc-AxhUy jXKcLv"
      >
        <div
          class="sc-AxheI hujUmw"
        >
          <label
            class="sc-fzplWN cnlTHl"
            data-testid="main-create"
          >
            <form
              data-testid="main-create-form"
            >
              <input
                class="sc-AxgMl gEwrMB"
                data-testid="main-create-form-input"
                value=""
              />
            </form>
            <button
              class="sc-AxjAm RHRov"
              data-testid="main-create-button"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="12"
                  x2="12"
                  y1="5"
                  y2="19"
                />
                <line
                  x1="5"
                  x2="19"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
          </label>
          <div
            class="sc-AxheI ghkIZM"
          >
            <div
              class="sc-Axmtr cassBX"
              data-testid="todo"
            >
              <div
                class="sc-AxmLO QPZRk"
                data-testid="todo-completed"
              >
                <label
                  class="sc-AxirZ ATkXc"
                >
                  <input
                    checked=""
                    class="sc-AxhCb iKostn"
                    data-testid="todo-completed-input"
                    type="checkbox"
                  />
                  <div
                    class="sc-AxiKw CKSLb"
                  >
                    <svg
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <polyline
                        points="20 6 9 17 4 12"
                      />
                    </svg>
                  </div>
                </label>
              </div>
              <div
                class="sc-fzozJi eXhrpV"
                data-testid="todo-content"
              >
                <input
                  class="sc-AxgMl gEwrMB"
                  data-testid="todo-content-input"
                  value="create todo app"
                />
              </div>
              <div
                class="sc-fzoLsD hWbsas"
              >
                <button
                  class="sc-AxjAm RHRov"
                  data-testid="todo-button"
                >
                  <svg
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <line
                      x1="18"
                      x2="6"
                      y1="6"
                      y2="18"
                    />
                    <line
                      x1="6"
                      x2="18"
                      y1="6"
                      y2="18"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              class="sc-Axmtr cassBX"
              data-testid="todo"
            >
              <div
                class="sc-AxmLO QPZRk"
                data-testid="todo-completed"
              >
                <label
                  class="sc-AxirZ bzNmNo"
                >
                  <input
                    class="sc-AxhCb iKostn"
                    data-testid="todo-completed-input"
                    type="checkbox"
                  />
                  <div
                    class="sc-AxiKw CKSLb"
                  >
                    <svg
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <polyline
                        points="20 6 9 17 4 12"
                      />
                    </svg>
                  </div>
                </label>
              </div>
              <div
                class="sc-fzozJi eXhrpV"
                data-testid="todo-content"
              >
                <input
                  class="sc-AxgMl gEwrMB"
                  data-testid="todo-content-input"
                  value="find new job"
                />
              </div>
              <div
                class="sc-fzoLsD hWbsas"
              >
                <button
                  class="sc-AxjAm RHRov"
                  data-testid="todo-button"
                >
                  <svg
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <line
                      x1="18"
                      x2="6"
                      y1="6"
                      y2="18"
                    />
                    <line
                      x1="6"
                      x2="18"
                      y1="6"
                      y2="18"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
          <div
            class="sc-fznKkj FGnGs"
            data-testid="main-options"
          >
            <button
              class="sc-AxjAm RHRov"
              data-testid="main-options-checkall"
            >
              Check All
            </button>
            <button
              class="sc-AxjAm RHRov"
              data-testid="main-options-uncheckall"
            >
              Uncheck All
            </button>
            <button
              class="sc-AxjAm RHRov"
              data-testid="main-options-deleteall"
            >
              Delete All
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-AxhUy jXKcLv"
    >
      <div
        class="sc-AxheI hujUmw"
      >
        <label
          class="sc-fzplWN cnlTHl"
          data-testid="main-create"
        >
          <form
            data-testid="main-create-form"
          >
            <input
              class="sc-AxgMl gEwrMB"
              data-testid="main-create-form-input"
              value=""
            />
          </form>
          <button
            class="sc-AxjAm RHRov"
            data-testid="main-create-button"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="12"
                x2="12"
                y1="5"
                y2="19"
              />
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
            </svg>
          </button>
        </label>
        <div
          class="sc-AxheI ghkIZM"
        >
          <div
            class="sc-Axmtr cassBX"
            data-testid="todo"
          >
            <div
              class="sc-AxmLO QPZRk"
              data-testid="todo-completed"
            >
              <label
                class="sc-AxirZ ATkXc"
              >
                <input
                  checked=""
                  class="sc-AxhCb iKostn"
                  data-testid="todo-completed-input"
                  type="checkbox"
                />
                <div
                  class="sc-AxiKw CKSLb"
                >
                  <svg
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <polyline
                      points="20 6 9 17 4 12"
                    />
                  </svg>
                </div>
              </label>
            </div>
            <div
              class="sc-fzozJi eXhrpV"
              data-testid="todo-content"
            >
              <input
                class="sc-AxgMl gEwrMB"
                data-testid="todo-content-input"
                value="create todo app"
              />
            </div>
            <div
              class="sc-fzoLsD hWbsas"
            >
              <button
                class="sc-AxjAm RHRov"
                data-testid="todo-button"
              >
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="18"
                    x2="6"
                    y1="6"
                    y2="18"
                  />
                  <line
                    x1="6"
                    x2="18"
                    y1="6"
                    y2="18"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="sc-Axmtr cassBX"
            data-testid="todo"
          >
            <div
              class="sc-AxmLO QPZRk"
              data-testid="todo-completed"
            >
              <label
                class="sc-AxirZ bzNmNo"
              >
                <input
                  class="sc-AxhCb iKostn"
                  data-testid="todo-completed-input"
                  type="checkbox"
                />
                <div
                  class="sc-AxiKw CKSLb"
                >
                  <svg
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <polyline
                      points="20 6 9 17 4 12"
                    />
                  </svg>
                </div>
              </label>
            </div>
            <div
              class="sc-fzozJi eXhrpV"
              data-testid="todo-content"
            >
              <input
                class="sc-AxgMl gEwrMB"
                data-testid="todo-content-input"
                value="find new job"
              />
            </div>
            <div
              class="sc-fzoLsD hWbsas"
            >
              <button
                class="sc-AxjAm RHRov"
                data-testid="todo-button"
              >
                <svg
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="18"
                    x2="6"
                    y1="6"
                    y2="18"
                  />
                  <line
                    x1="6"
                    x2="18"
                    y1="6"
                    y2="18"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="sc-fznKkj FGnGs"
          data-testid="main-options"
        >
          <button
            class="sc-AxjAm RHRov"
            data-testid="main-options-checkall"
          >
            Check All
          </button>
          <button
            class="sc-AxjAm RHRov"
            data-testid="main-options-uncheckall"
          >
            Uncheck All
          </button>
          <button
            class="sc-AxjAm RHRov"
            data-testid="main-options-deleteall"
          >
            Delete All
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
